Stripe more characters from style settings

Crabcyborg opened the pull request:
Related issue https://github.com/Strategy11/formidable-pro/issues/3830

I'm not really sure how Nat's CSS is the way it is, but I think it could just be a weird setting somewhere.

It's been an ongoing project to try cleaning up our CSS, like our rgba color normalizing.

It looks like `{`, `}`, ';' are never stripped so if you accidentally put one in the CSS it will break the file.

@stephywells Do you see any of these characters being required for anything? Do you think there are other characters or phrases we'd want to remove as well?

stephywells reviewed (approved):
That looks good. This doesn't apply to the custom css box, right?

Crabcyborg commented:
> That looks good. This doesn't apply to the custom css box, right?

@stephywells It doesn't since custom CSS isn't part of the style post content anymore.

But I think there could be a risk about old custom CSS that hasn't yet migrated getting changed. I added a check for the `custom_css` key just in case as we wouldn't want it to ever happen for custom CSS.

Thanks Steph!

garretlaxton commented:
What about stripping these as well? `(` `)` `[` `]`. Or maybe just creating an allow list of what we expect to be used in the CSS instead of trying to strip certain characters?

I was able to break the styling again with those.
![image](https://user-images.githubusercontent.com/16290508/230477099-a9ee54d6-cfd4-40a5-b664-9d139f0721ff.png)

Crabcyborg commented:
Thanks @garretlaxton!

I don't believe we ever expect `[`/`]` braces so I'm removing those now too.

As for the `(`/`)` braces, those are used in RGB(A) colors and also in CSS like `calc( 100% / 3 )`. So what I did was added a check - first that it ignores RGBA (as that is handled differently), and so that it only removes `(`/`)` when the number doesn't match up. So `calc( 100% / 3 )` is left alone but if you do something like `fff)` or `000(` those extra braces will be stripped.

stephywells commented:
@Crabcyborg we could try adding an end ) if there isn't one, like the WP force_balance_tags() function. That one only covered HTML though, but I just mean the idea of it.

Crabcyborg commented:
@stephywells Maybe that would be useful in some cases, like if it's a `calc` but I feel like you're more likely to hit an accidental character. `fff(` is better as `fff` than it is as `fff()` which would have issues too, so there are trade-offs.

garretlaxton reviewed (approved):
Nice, this looks good now!

Crabcyborg commented:
@stephywells I've pushed a pretty large update to this to handle `(` and `)` more intentionally.

I try to strip all `(` and `)` now from a value that is detected as a size or a hex with extra `(` or `)` characters.

I considered a few other things like checking for `calc`, `var`, etc but the list of possible CSS functions is quite large (it looks like there may be around 48 and likely more in the future) like `skewX`, `translateX`, `scaleX`, `var`, etc. So I abandoned that idea.

I also trim excessive `(` and `)` when they are expected. So `(calc(100%))` becomes `calc(100%)` and `skewX(5px)))` becomes `skewX(5px)`, etc.

I also updated `force_balanced_quotation` as it didn't really work as expected. I added unit tests for it and noticed `Arial"` for example would get "fixed" to `Arial""` when `"Arial"` makes a lot more sense.

stephywells reviewed (commented):
Those are some good points. I think a basic pass should be fine. One thing we could do is add a .frm_css_error{display:none} at the very end of the css file.

Then, if there is any issue with the css, an error inside that close would display. This could tell people what to look for in their settings, even if we don't fix it automatically. Right now, it takes time to find that the css is broken.

Crabcyborg commented on classes/models/FrmStyle.php:
```diff
@@ -241,10 +241,101 @@ public function sanitize_post_content( $settings ) {
+		}
+
+		// Matches size values but also checks for unexpected ( and ).
+		$looks_like_a_size = preg_match( '/\(?[+-]?\d*\.?\d+(?:px|%|em|rem|ex|pt|pc|mm|cm|in)\)?/', $setting );
```
Good call @stephywells! I've updated this to make it case insensitive and added an assert for a "PX" variation.

Crabcyborg commented:
> One thing we could do is add a .frm_css_error{display:none} at the very end of the css file.

I haven't done this yet. I'll play with this next.

I've pushed a new update that tackles some other things though,

- I've updated the brace trimming so `)` gets removed from the left and `(` gets removed from the right as well. The previous version just checked for extra of the expected type of brace on either side.
- The sizes are now case insensitive based on this comment https://github.com/Strategy11/formidable-forms/pull/1141#discussion_r1163279459
- Added some asserts for calc() rules.

I also considered other CSS values like for standard colors like `red(`, and `)blue` but realized the color picker doesn't support those so it isn't necessary to check for those.

stephywells commented:
That sounds great! Thanks for thinking through so many possibilities!
